pallets/msa/src/lib.rs

```diff
@@ -1129,6 +1134,49 @@ impl<T: Config> Pallet<T> {
 		})
 	}
 
+	/// Modify delegation's schema permissions
+	///
+	/// # Errors
+	/// * [`Error::ExceedsMaxSchemaGrantsPerDelegation`]
+	pub fn upsert_schema_permissions(
+		provider_id: ProviderId,
+		delegator_id: DelegatorId,
+		schema_ids: Vec<SchemaId>,
+	) -> DispatchResult {
+		Self::try_mutate_delegation(delegator_id, provider_id, |delegation, _is_new_delegation| {
+			Self::ensure_all_schema_ids_are_valid(&schema_ids)?;
+
+			// Create revoke and insert lists
+			let mut revoke_ids: Vec<SchemaId> = Vec::new();
+			let mut insert_ids: Vec<SchemaId> = Vec::new();
+
+			let existing_keys: Vec<SchemaId> =
+				delegation.schema_permissions.keys().cloned().collect();
+			for existing_schema_id in &existing_keys {
```

Comment:
I wonder if using a hashset would be worth the cost here. If we assume that most users have 10 schemas (in and out), this is ~`O(10*5*2) = 100` between these two nested contains checks. With the worse case being `O(30*30*2) = 1_800`  (with MaxSchemas set to 30 and users swapping 30 for 30).

Assuming a set is possible (with normal insert at `O(1)` and check at `O(1)`) then 10 worst case would be `O(10+10+10+10) = 40` and 30 worst case would be `O(30+30+30+30) = 120`

Something like tinyset: https://crates.io/crates/tinyset would work with default features disabled.
hashbrown is already used in substrate, but not in a no_std env as far as I can tell, but might work: https://github.com/rust-lang/hashbrown#features

Reply:
I think we should consider it in a future story.

Reply:
Agreed. Don't want it to block the next release. I'll add it to the top of the tech debt so it gets done soon.

Reply:
Added an updated version of the suggestion into a new story: https://github.com/LibertyDSNP/frequency/issues/1155